#VRML_SIM R2023b utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/robots/gctronic/e-puck/protos/E-puckDistanceSensor.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/robots/nvidia/jetbot/protos/JetBotRaspberryPiCamera.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2023b/projects/robots/mir/mir100/protos/Mir100UltrasonicSensor.proto"

PROTO AutoNova [
  field SFVec3f            translation                  0 0 0
  field SFRotation         rotation                     0 0 1 0
  field SFString           name                         "AutoNova"
  field SFString           controller                   "<generic>"
]
{
Robot {
  translation IS translation
  rotation IS rotation
  children [
    GPS {
      name "GPS"
    }
    Mir100UltrasonicSensor {
      translation 0 -0.073 0.009
      rotation 0 0 1 -1.5708
      name "ps0"
    }
    Mir100UltrasonicSensor {
      translation 0 0.073 0.009
      rotation 0 0 1 1.5708
      name "ps4"
    }
    Mir100UltrasonicSensor {
      translation 0.073 0 0.009
      
      name "ps2"
    }
    Mir100UltrasonicSensor {
      translation -0.073 0 0.009
      rotation 0 0 1 3.14159
      name "ps8"
    }
    Gyro {
      name "gyroScope"
    }
    InertialUnit {
      name "Inertial Unit"
    }
    SpotLight {
      attenuation 0 0 0.4
      beamWidth 0.7
      direction 1 0 -0.4
      intensity 0
    }
    JetBotRaspberryPiCamera {
      translation 0.075 0 0.037
      fieldOfView 1
      width 360
      height 360
    }
    Mir100UltrasonicSensor {
      translation -0.050104 0.050104 0.038
      rotation 0 0 1 1.57  # Adjusted to align with the diagonal top-left
      name "ps7"
    }

    Mir100UltrasonicSensor {
      translation -0.050104 -0.050104 0.038
      rotation  0 0 1 -1.57  # Adjusted to align with bottom-left
      name "ps5"
    }

    Mir100UltrasonicSensor {
      translation 0.050104 -0.050104 0.038
      rotation 0 0 1 -1.57  # Adjusted to align with bottom-right
      name "ps3"
    }

    Mir100UltrasonicSensor {
      translation 0.050104 0.050104 0.038
      rotation 0 0 1 1.57  # Adjusted to align with top-right
      name "ps1"
    }

    
    HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor 0 0.06 0.0098
      }
      device [
        PositionSensor {
          name "left encoder"
        }
        RotationalMotor {
          name "left motor"
        }
      ]
      endPoint Solid {
        translation 0 0.06 0.0098
        rotation 1 -6.180011507259454e-11 6.180033097574564e-11 1.5707999999974487
        children [
          DEF wheel Shape {
            appearance PBRAppearance {
              baseColor 0 1 1
              metalness 0
            }
            geometry DEF wheel_geo Cylinder {
              height 0.009
              radius 0.03
            }
          }
        ]
        name "LeftWheel"
        boundingObject USE wheel
        physics Physics {
        }
      }
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        axis 0 1 0
        anchor 0 -0.06 0.0098
      }
      device [
        PositionSensor {
          name "right encoder"
        }
        RotationalMotor {
          name "right motor"
        }
      ]
      endPoint Solid {
        translation 0 -0.06 0.0098
        rotation 1 -6.21802815342693e-11 6.21805007065936e-11 1.5707999999974478
        children [
          DEF wheel Shape {
            appearance PBRAppearance {
              baseColor 0 1 1
              metalness 0
            }
            geometry DEF wheel_geo Cylinder {
              height 0.009
              radius 0.03
            }
          }
        ]
        name "RightWheel"
        boundingObject USE wheel
        physics DEF physics_wh Physics {
        }
      }
    }
    DEF Body Shape {
      appearance PBRAppearance {
        metalness 0
      }
      geometry Box {
        size 0.1 0.1 0.0399
      }
    }
    Solid {
      translation -0.0859 -0.0853 0.02
      children [
        Shape {
          appearance PBRAppearance {
            metalness 0
          }
          geometry Mesh {
            url [
              "../RoboParts/topCylinder.STL"
            ]
            name "top cylinder"
          }
        }
      ]
      name "top"
      boundingObject Mesh {
        url [
          "../RoboParts/topCylinder.STL"
        ]
      }
      physics Physics {
        density 0.1
        centerOfMass [
          0 0 0
        ]
      }
    }
  ]
  name "AutoNova"
  model "v1"
  boundingObject USE Body
  physics Physics {
  }
  controller IS controller
  battery [
    0
  ]
}
}

